src/ape/pytest/contextmanagers.py

```diff
@@ -1,26 +1,141 @@
+            raise AssertionError(
+                "Cannot check dev message; provider must support transaction tracing."
+            )
+
+        if trace[-1].op != "REVERT":
+            raise AssertionError("Transaction trace did not end with REVERT.")
+
+        pc = None
+        pcmap = contract.contract_type.pcmap.parse()
+
+        # To find a suitable line for inspecting dev messages, we must start at the revert and work
+        # our way backwards. If the last frame's PC is in the PC map, the offending line is very
+        # likely a 'raise' statement.
+        if trace[-1].pc in contract.contract_type.pcmap:
+            pc = trace[-1].pc
+
+        # Otherwise we must traverse the trace backwards until we find our first suitable candidate.
+        # It should be noted that when an error happens, typically the contract will execute:
+        #
+        #   <...>
+        #   JUMPI    <- jump to revert routine on a failing condition
+        #
+        #   <...>
+        #   JUMPDEST <- beginning of revert routine
+        #   PUSH1
+        #   DUP1
+        #   REVERT
+        #
+        # So we must step backwards until we find the initial instructions that caused the failing
+        # condition. This will not be the JUMPI or even an arithmetic instruction like DIV, but
+        # rather the starting instruction for a given line.
+        else:
+            while len(trace) > 0:
+                frame = trace.pop()
+                if frame.pc in pcmap:
+                    pc = frame.pc
+                    break
+
+        # We were unable to find a suitable PC that matched the compiler's map.
+        if pc is None:
+            raise AssertionError("Could not find line that caused revert.")
+
+        offending_source = pcmap[pc]
+
+        # The compiler PC map had PC information, but not source information.
+        if offending_source is None or offending_source.line_start is None:
+            raise AssertionError("Could not find line that caused revert.")
+
+        try:
+            contract_source_lines = (
+                ape.project.extract_manifest()
+                .sources[contract.contract_type.source_id]
+                .content.split("\n")
+            )
+        except KeyError as exc:
+            raise AssertionError("Could not retrieve contract source.") from exc
+
+        if offending_line.line_start > len(contract_source_lines):
+            raise AssertionError(
+                "Could not find line that caused revert (line outside of contract range)."
+            )
 
-        actual = exc_value.revert_message
+        # Line numbers will not be zero-indexed
+        offending_line = contract_source_lines[offending_source.line_start - 1]
+
+        # Match '# dev: (...)' with any variation of whitespace
+        match = re.search(r"#\s*dev\s*:\s*(.+)", offending_line)
+
+        if not match:
+            raise AssertionError(f"Could not find dev message '{self.dev_message}'.")
```

Comment:
@NotPeopling2day @unparalleled-js @fubuloubu thoughts on the error messaging - should we print out the offending line (up to a certain number of characters perhaps)?

Reply:
Sure! That sounds like it goes with our initiative to show traces in place of python exceptions during uncaught reverts.